rr_iarrc/src/finish_line_watcher/finish_line_watcher.cpp

```diff
@@ -78,30 +73,118 @@ void ImageCB(const sensor_msgs::ImageConstPtr& msg) {
     frame = cv_ptr->image;
     blockEnvironment(frame);
 
-    auto count = countNonZero(frame);
+    // Matrix used to draw debug info
+    cv::Mat debugDrawing = cv::Mat::zeros(frame.size(), CV_8UC3);
+    cv::cvtColor(frame, debugDrawing, cv::COLOR_GRAY2BGR);
+
+    // Find contours
+    vector<vector<cv::Point>> contours;
+    cv::findContours(frame, contours, cv::RETR_LIST, cv::CHAIN_APPROX_SIMPLE);
+
+    // Filter out very small contours
+    vector<cv::Point> filteredContourPoints;
+    for (int i = 0; i < contours.size(); i++) {
+        cv::Scalar color(0, 0, 255);
 
-    auto width = getWidth(frame);
+        if (cv::contourArea(contours.at(i)) > min_contour_area) {
+            filteredContourPoints.insert(std::end(filteredContourPoints), std::begin(contours.at(i)),
+                                         std::end(contours.at(i)));
+            color = cv::Scalar(0, 255, 0);
+        }
+
+        // Debug draw
+        cv::drawContours(debugDrawing, contours, i, color, 2, 8);
+    }
 
-    if (state == LOW && count > 2000 && width > 700) {
+    bool detected = false;
+    bool drawRectDebug = false;
+    cv::RotatedRect fitRect;
+    cv::Size2f size;
+    float angle;
+
+    if (filteredContourPoints.size() > 0) {
+        // Fit a rect to the points and detect based on angle and width
+        fitRect = cv::minAreaRect(filteredContourPoints);
+        size = fitRect.size;
+        drawRectDebug = true;
+        // Check the detection criteria (must look at both width and height because OpenCV is inconsistent about how it
+        // assigns them)
+        if (size.height > size.width)
+            angle = std::abs(fitRect.angle) - 90;
```

Comment:
Where do you update `angle` when this is skipped? I think you are leaving it un-initialized

Reply:
That should be the last fix